### PR TITLE
[zh-cn]: update the translation of for...of... in statement section

### PR DESCRIPTION
**Pull request by JasonLamv-t**
### Description

As the title.

### Motivation

Ditto.

### Additional details

None.

### Related issues and pull requests

None.

### CODE REVIEW

File: files/zh-cn/web/javascript/reference/statements/for...of/index.md [hunk 1]
```diff
@@ -1,33 +1,46 @@
...
+
+当 `for...of` 循环迭代一个可迭代对象时，它首先调用可迭代对象的 [`[@@iterator]()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator) 方法，该方法返回一个[迭代器](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#迭代器协议)，然后重复调用生成器的 [`next()`](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#迭代器协议) 方法，以生成要分配给 `variable` 的值的序列。
 
-### 迭代{{jsxref("Array")}}
+`for...of` 循环在迭代器完成时退出（即迭代器的 `next()` 方法返回一个包含 `done: true` 的对象）。你也可以使用控制流语句来改变正常的控制流程。[`break`](/zh-CN/docs/Web/JavaScript/Reference/Statements/break) 语句退出循环并跳转到循环体后的第一条语句，而 [`continue`](/zh-CN/docs/Web/JavaScript/Reference/Statements/continue) 语句跳过当前迭代的剩余语句，继续进行下一次迭代。
+
+如果 `for...of` 循环提前退出（例如遇到 `break` 语句或抛出错误），则会调用迭代器的 [`return()`](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#迭代器协议) 方法来执行任何清理任务。
+
+The `variable` part of `for...of` accepts anything that can come before the `=` operator. You can use {{jsxref("Statements/const", "const")}} to declare the variable as long as it's not reassigned within the loop body (it can change between iterations, because those are two separate variables). Otherwise, you can use {{jsxref("Statements/let", "let")}}.
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/statements/for...of/index.md:38]:
```suggestion
```

File: files/zh-cn/web/javascript/reference/statements/for...of/index.md [hunk 2]
```diff
@@ -1,33 +1,45 @@
...
+`for...of` 循环按顺序逐个处理从可迭代对象获取的值。循环对值的每次操作被称为一次*迭代*，而循环本身被称为*迭代可迭代对象*。每次迭代都会执行可能引用当前序列值的语句。
+
+当 `for...of` 循环迭代一个可迭代对象时，它首先调用可迭代对象的 [`[@@iterator]()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator) 方法，该方法返回一个[迭代器](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#迭代器协议)，然后重复调用生成器的 [`next()`](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#迭代器协议) 方法，以生成要分配给 `variable` 的值的序列。
 
-### 迭代{{jsxref("Array")}}
+`for...of` 循环在迭代器完成时退出（即迭代器的 `next()` 方法返回一个包含 `done: true` 的对象）。你也可以使用控制流语句来改变正常的控制流程。[`break`](/zh-CN/docs/Web/JavaScript/Reference/Statements/break) 语句退出循环并跳转到循环体后的第一条语句，而 [`continue`](/zh-CN/docs/Web/JavaScript/Reference/Statements/continue) 语句跳过当前迭代的剩余语句，继续进行下一次迭代。
+
+如果 `for...of` 循环提前退出（例如遇到 `break` 语句或抛出错误），则会调用迭代器的 [`return()`](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#迭代器协议) 方法来执行任何清理任务。
+
+
```
Comment (JasonLamv-t) [inline files/zh-cn/web/javascript/reference/statements/for...of/index.md:38]:
```suggestion
```

File: files/zh-cn/web/javascript/reference/statements/for...of/index.md [hunk 3]
```diff
@@ -1,33 +1,44 @@
...
 title: for...of
 slug: Web/JavaScript/Reference/Statements/for...of
+l10n:
+  sourceCommit: a71b8929628a2187794754c202ad399fe357141b
 ---
 
 {{jsSidebar("Statements")}}
 
-**`for...of`语句**在[可迭代对象](/zh-CN/docs/Web/JavaScript/Guide/iterable)（包括 {{jsxref("Array")}}，{{jsxref("Map")}}，{{jsxref("Set")}}，{{jsxref("String")}}，{{jsxref("TypedArray")}}，[arguments](/zh-CN/docs/Web/JavaScript/Reference/Functions_and_function_scope/arguments) 对象等等）上创建一个迭代循环，调用自定义迭代钩子，并为每个不同属性的值执行语句
+**`for...of`** 语句执行的循环作用于从[可迭代对象](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#可迭代协议)中获取的值序列。可迭代对象包括内置对象的实例，例如 {{jsxref("Array")}}、{{jsxref("String")}}、{{jsxref("TypedArray")}}、{{jsxref("Map")}}、{{jsxref("Set")}}、{{domxref("NodeList")}}（以及其他 DOM 集合），还包括 {{jsxref("Functions/arguments", "arguments")}} 对象、由[生成器函数](/zh-CN/docs/Web/JavaScript/Reference/Statements/function*)生成的[生成器](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Generator)，以及用户定义的可迭代对象。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/statements/for...of/index.md:10]: I'm not sure if this modification is appropriate, just feel the original translation a bit difficult to understand.

```suggestion
**`for...of`** 语句执行一个循环，该循环处理来自[可迭代对象](/zh-CN/docs/Web/JavaScript/Reference/Iteration_protocols#可迭代协议)的值序列。可迭代对象包括内置对象的实例，例如 {{jsxref("Array")}}、{{jsxref("String")}}、{{jsxref("TypedArray")}}、{{jsxref("Map")}}、{{jsxref("Set")}}、{{domxref("NodeList")}}（以及其他 DOM 集合），还包括 {{jsxref("Functions/arguments", "arguments")}} 对象、由[生成器函数](/zh-CN/docs/Web/JavaScript/Reference/Statements/function*)生成的[生成器](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Generator)，以及用户定义的可迭代对象。
```

File: files/zh-cn/web/javascript/reference/statements/for...of/index.md [hunk 4]
```diff
@@ -38,10 +49,25 @@ for (let value of iterable) {
 // 31
 ```
 
-如果你不想修改语句块中的变量 , 也可以使用[`const`](/zh-CN/docs/Web/JavaScript/Reference/Statements/const)代替[`let`](/zh-CN/docs/Web/JavaScript/Reference/Statements/let)。
+> **备注：** 每次迭代都会创建一个新的变量。在循环体内部重新赋值变量不会影响可迭代对象（在本例中，是一个数组）的原始值。
+
+你可以使用[解构赋值](/zh-CN/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)来分配多个局部变量，或者使用属性访问器（如 `for (x.y of iterable)`）来給对象属性赋值。
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/statements/for...of/index.md:54]: nit

```suggestion
你可以使用[解构赋值](/zh-CN/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)来分配多个局部变量，或者使用属性访问器（如 `for (x.y of iterable)`）来给对象属性赋值。
```

File: files/zh-cn/web/javascript/reference/statements/for...of/index.md [hunk 5]
```diff
@@ -51,249 +77,287 @@ for (const value of iterable) {
...
-```js
-for (let i in iterable) {
-  console.log(i); // logs 0, 1, 2, "foo", "arrCustom", "objCustom"
+for (const i in iterable) {
+  console.log(i);
 }
-```
-
-此循环仅以原始插入顺序记录`iterable` 对象的[可枚举属性](/zh-CN/docs/Web/JavaScript/Enumerability_and_ownership_of_properties)。它不记录数组**元素** `3`, `5`, `7` 或`hello`，因为这些**不是**枚举属性。但是它记录了数组**索引**以及`arrCustom`和`objCustom`。如果你不知道为什么这些属性被迭代，{{jsxref("Statements/for...in", "array iteration and for...in", "#Array_iteration_and_for...in")}}中有更多解释。
+// "0", "1", "2", "foo", "arrCustom", "objCustom"
```
Comment (yin1999) [inline files/zh-cn/web/javascript/reference/statements/for...of/index.md:339]:
```suggestion
// "0"、"1"、"2"、"foo"、"arrCustom"、"objCustom"
```

**Review by yin1999 — APPROVED** [review]
Thank you @JasonLamv-t

**Pull request merged**